### ISSUE TITLE
Server crash on when File not found

### ISSUE THREAD

**Issue by PackeTsar** [msoulier/tftpy#79]
Getting the below error when client asks for a file which doesn't exist.

Looks like tftpy sends a proper error message and kills the session, but then it receives another packet (maybe an error ack?) and that causes it to crash.

```
[root@CENT7 tftpy]# tftpy_server.py -r "/" -d
DEBUG:tftpy:tftproot / does exist
DEBUG:tftpy:tftproot / is a directory
DEBUG:tftpy:tftproot / is readable
DEBUG:tftpy:tftproot / is writable
INFO:tftpy:Server requested on ip 0.0.0.0, port 69
INFO:tftpy:Starting receive loop...
DEBUG:tftpy:shutdown_immediately is False
DEBUG:tftpy:shutdown_gracefully is False
DEBUG:tftpy:Performing select on this inputlist: [<socket._socketobject object at 0x7f78d452de50>]
DEBUG:tftpy:Data ready on our main socket
DEBUG:tftpy:Read 13 bytes
DEBUG:tftpy:Creating new server context for session key = 10.162.0.2:58833
DEBUG:tftpy:In TftpContextServer.start
DEBUG:tftpy:Set metrics.start_time to 1494894666.47
DEBUG:tftpy:parsing a 13 byte packet
DEBUG:tftpy:opcode is 1
DEBUG:tftpy:in TftpPacketWithOptions.setoptions
DEBUG:tftpy:options: {}
DEBUG:tftpy:setting options hash to: {}
DEBUG:tftpy:in decode: about to iterate buffer counting nulls
DEBUG:tftpy:found a null at length 4, now have 1
DEBUG:tftpy:found a null at length 5, now have 2
DEBUG:tftpy:hopefully found end of mode at length 10
DEBUG:tftpy:about to unpack buffer with format: 4sx5sx
DEBUG:tftpy:unpacking buffer: 'some\x00octet\x00'
DEBUG:tftpy:set filename to some
DEBUG:tftpy:set mode to octet
DEBUG:tftpy:decode_options: buffer is: ''
DEBUG:tftpy:size of buffer is 0 bytes
DEBUG:tftpy:size of buffer is zero, returning empty hash
DEBUG:tftpy:in TftpPacketWithOptions.setoptions
DEBUG:tftpy:options: {}
DEBUG:tftpy:setting options hash to: {}
DEBUG:tftpy:in TftpPacketWithOptions.getoptions
DEBUG:tftpy:TftpContextServer.start() - factory returned a RRQ packet: filename = some mode = octet
DEBUG:tftpy:In TftpStateServerStart.handle
DEBUG:tftpy:Handling an RRQ packet
DEBUG:tftpy:In TftpStateServerRecvRRQ.handle
DEBUG:tftpy:in TftpPacketWithOptions.getoptions
INFO:tftpy:Setting tidport to 58833
DEBUG:tftpy:Setting default options, blksize
DEBUG:tftpy:Requested filename is some
DEBUG:tftpy:full_path is /some
INFO:tftpy:requested file is in the server root - good
INFO:tftpy:Opening file /some for reading
DEBUG:tftpy:In sendError, being asked to send error 1
58833
<type 'int'>
DEBUG:tftpy:encoding ERR packet with format !HH14sx
ERROR:tftpy:Fatal exception thrown from session 10.162.0.2:58833: File not found: /some
INFO:tftpy:Currently handling these sessions:
INFO:tftpy:    10.162.0.2:58833 <tftpy.TftpStates.TftpStateServerStart object at 0x7f78d4549590>
DEBUG:tftpy:Looping on all sessions to check for timeouts
DEBUG:tftpy:checking for timeout on session 10.162.0.2:58833 <tftpy.TftpStates.TftpStateServerStart object at 0x7f78d4549590>
DEBUG:tftpy:Iterating deletion list.
INFO:tftpy:
INFO:tftpy:Session 10.162.0.2:58833 complete
DEBUG:tftpy:Gathering up metrics from session before deleting
DEBUG:tftpy:in TftpContext.end
DEBUG:tftpy:Set metrics.end_time to 1494894666.47
DEBUG:tftpy:TftpMetrics.compute: duration is 0.00209307670593
DEBUG:tftpy:TftpMetrics.compute: kbps is 0.0
INFO:tftpy:Transferred 0 bytes in 0.00 seconds
INFO:tftpy:Average rate: 0.00 kbps
INFO:tftpy:0.00 bytes in resent data
INFO:tftpy:0 duplicate packets
DEBUG:tftpy:Deleting session 10.162.0.2:58833
DEBUG:tftpy:Session list is now {}
DEBUG:tftpy:shutdown_immediately is False
DEBUG:tftpy:shutdown_gracefully is False
DEBUG:tftpy:Performing select on this inputlist: [<socket._socketobject object at 0x7f78d452de50>]
DEBUG:tftpy:Data ready on our main socket
DEBUG:tftpy:Read 23 bytes
DEBUG:tftpy:Creating new server context for session key = 10.162.0.2:58833
DEBUG:tftpy:In TftpContextServer.start
DEBUG:tftpy:Set metrics.start_time to 1494894666.47
DEBUG:tftpy:parsing a 23 byte packet
DEBUG:tftpy:opcode is 5
DEBUG:tftpy:Decoding ERR packet, length 23 bytes
DEBUG:tftpy:Good ERR packet > 4 bytes
DEBUG:tftpy:Decoding ERR packet with format: !HH18sx
ERROR:tftpy:ERR packet - errorcode: 0, message: Session terminated
DEBUG:tftpy:TftpContextServer.start() - factory returned a ERR packet: errorcode = 0
    msg = 
DEBUG:tftpy:In TftpStateServerStart.handle
DEBUG:tftpy:In sendError, being asked to send error 4
None
<type 'NoneType'>
DEBUG:tftpy:encoding ERR packet with format !HH22sx
Traceback (most recent call last):
  File "/usr/bin/tftpy_server.py", line 59, in <module>
    main()
  File "/usr/bin/tftpy_server.py", line 51, in main
    server.listen(options.ip, options.port)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpServer.py", line 166, in listen
    self.sessions[key].start(buffer)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpContexts.py", line 241, in start
    self.port)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 423, in handle
    self.sendError(TftpErrors.IllegalTftpOp)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 133, in sendError
    self.context.tidport))
TypeError: an integer is required
```

**Comment by PackeTsar** [comment]
Looks like the packet sent back from the client (which causes the crash) is a opcode 5, error code 0, message "Session terminated". Packet capture attached if it helps.
[tftp.pcap.zip](https://github.com/msoulier/tftpy/files/1002849/tftp.pcap.zip)

**Comment by PackeTsar** [comment]
I've confirmed that the issue is a third packet sent only by certain platforms:

The "Good Fail" - Where tftpy responds with a "File Not Found" and doesn't crash had a Cisco 2800 router as a client

The "Bad Fail" - Where tftpy raises a TypeError exception and crashes had a Cisco 3750 switch as a client.
[Archive.zip](https://github.com/msoulier/tftpy/files/1055843/Archive.zip)

**Comment by PackeTsar** [comment]
Fixed in a fork. Submitting a PR for this and issue #78

**Comment by msoulier** [comment]
I can't repro this on Linux, but I did find another problem. My server is not responding with file not found at all. It claims that the file exists and then sends a 0 size DAT packet and transfers a 0 size file.

**Comment by msoulier** [comment]
I moved my server root to /tmp and requested "nosuchfile", and it worked fine.

```
[2018-05-15 12:10:00,652652] DEBUG [tftpy.TftpStates:129] In sendError, being asked to send error 4
[2018-05-15 12:10:00,652652] DEBUG [tftpy.TftpPacketTypes:380] encoding ERR packet with format !HH22sx
Received ERR from server: ERR packet: errorcode = 1
    msg = File not found
```

**Comment by msoulier** [comment]
Mind you it still created an empty "nosuchfile" as output. That's broken.

**Comment by msoulier** [comment]
So, TftpContextClientDownload is opening the output file. We really should delete it on a file not found.

**Comment by PackeTsar** [comment]
Did you see the packet capture where the client sends a third error packet and crashes the server?

**Comment by msoulier** [comment]
Ok, fixed that by raising a new TftpFileNotFoundError exception from the state handler, and the context can now intercept and unlink the file, if it's a file.

**Comment by msoulier** [comment]
Yeah, errorcode 0 likely won't go over well. I should add one of those to the unit tests.

**Comment by msoulier** [comment]
Should be fixed in 45694f2902af7472fea903d3c6e71a637bc10e18.
